pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -876,8 +868,28 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env *model.E
 						Service:                    service,
 					}
 
-					configgen.buildSidecarOutboundListenerForPortOrUDS(node, listenerOpts, pluginParams, listenerMap,
-						virtualServices, actualWildcard)
+					// Minor optimization for Kubernetes statefulsets/headless services
+					// Instead of generating a single 0.0.0.0:Port listener, generate a listener
+					// for each podIP.
+					if bind == "" && service.Resolution == model.Passthrough &&
+						service.Attributes.ServiceRegistry == string(serviceregistry.KubernetesRegistry) {
+						if instances, err := env.InstancesByPort(service, servicePort.Port, nil); err == nil {
+							for _, instance := range instances {
+								listenerOpts.bind = instance.Endpoint.Address
+								configgen.buildSidecarOutboundListenerForPortOrUDS(node, listenerOpts, pluginParams, listenerMap,
+									virtualServices, actualWildcard)
+							}
+						} else {
+							// we can't do anything. Fallback to the usual way of constructing listeners
```

Comment:
when does this happen?

Reply:
Normally this can not happen, just in case there is an error returned.

Reply:
if env.InstancesByPort fails.